<a href="https://colab.research.google.com/github/eduardopadilla3/DS-Unit-1-Sprint-2-Statistics/blob/master/%5Cmodule1%5CCopy_of_LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [0]:
### YOUR WORK HERE
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt 
import seaborn as sns
from scipy import stats

In [0]:
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data'

In [0]:
df = pd.read_csv(url)

In [0]:
columns = ['symboling','norm_loss','make','fuel','aspiration','doors',
           'bod_style','drv_wheels','eng_loc','wheel_base','length','width',
           'height','curb_weight','engine','cylinders','engine_size',
           'fuel_system','bore','stroke','compression','hp','peak_rpm',
           'city_mpg','hgwy_mpg','price']

In [0]:
df = pd.read_csv(url, header=0, names=columns)
df.head()

symboling norm_loss         make fuel  ... peak_rpm city_mpg hgwy_mpg  price
0          3         ?  alfa-romero  gas  ...     5000       21       27  16500
1          1         ?  alfa-romero  gas  ...     5000       19       26  16500
2          2       164         audi  gas  ...     5500       24       30  13950
3          2       164         audi  gas  ...     5500       18       22  17450
4          2         ?         audi  gas  ...     5500       19       25  15250

[5 rows x 26 columns]

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and compare your sample mean estimate to that value. Can you say that they are different? 

In [0]:
### YOUR WORK HERE
sample1 = df.sample(10, random_state=30)


In [0]:
sample1

symboling norm_loss           make  ... city_mpg hgwy_mpg  price
146          0        89         subaru  ...       25       31  10198
67          -1        93  mercedes-benz  ...       22       25  28248
26           1       148          dodge  ...       24       30   8558
155          0        91         toyota  ...       30       37   6938
135          3       150           saab  ...       19       26  18150
24           1       148          dodge  ...       31       38   6692
154          0        91         toyota  ...       27       32   8778
195         -2       103          volvo  ...       24       28  15985
132          2       104           saab  ...       21       28  12170
194         -1        74          volvo  ...       23       28  13415

[10 rows x 26 columns]

In [0]:
df.hgwy_mpg.mean()

30.769607843137255

In [0]:
sample1.hgwy_mpg.mean()

30.3

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

In [0]:
### YOUR WORK HERE


In [0]:
stats.ttest_1samp(sample1['hgwy_mpg'], 35, nan_policy='omit')

Ttest_1sampResult(statistic=-3.4187469470305465, pvalue=0.007643069993182772)

Based on a t-statistic of -3.4 and a p-value of .0076, I reject the null hypothesis that congressional support for the handicapped-infants bill is equal to 50%.

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, what would you conclude when sampling 100 cars?

In [0]:
### YOUR WORK HERE
sample2 = df.sample(100, random_state=30)


In [0]:
stats.ttest_1samp(sample2['hgwy_mpg'], 35, nan_policy='omit')

Ttest_1sampResult(statistic=-6.4827186446460106, pvalue=3.5481090083532426e-09)

Based on a t-statistic of -6.48 and a p-value of 0, I reject the null hypothesis that congressional support for the handicapped-infants bill is equal to 50%.






Data Science Unit 1 Sprint 2 Assignment 1
Apply the t-test to real data
Practice 1-Sample T-tests
One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean.

1) Load the Data
Use the automobile dataset
Fix the column headers
Make sure NaNs are used to indicate missing values
Feel free to add code cells and text cells as needed throughout the assignment.

[47]
### YOUR WORK HERE
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt 
import seaborn as sns
from scipy import stats
[48]
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data'
[49]
df = pd.read_csv(url)
[50]
columns = ['symboling','norm_loss','make','fuel','aspiration','doors',
           'bod_style','drv_wheels','eng_loc','wheel_base','length','width',
           'height','curb_weight','engine','cylinders','engine_size',
           'fuel_system','bore','stroke','compression','hp','peak_rpm',
           'city_mpg','hgwy_mpg','price']
[51]
df = pd.read_csv(url, header=0, names=columns)
df.head()

2) Pretend that this dataset represents the cars at a used car lot in your local town.
Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the random_state so that all of us in the class get the same random sample. Please set your random state to 30 when using df.sample()
Is your sample reflective of the population value in regards to highway-mpg? Find the mean for highway-mpg for the entire dataset and compare your sample mean estimate to that value. Can you say that they are different?

[ ]
### YOUR WORK HERE
sample1 = df.sample(10, random_state=30)

[ ]
sample1

[ ]
df.hgwy_mpg.mean()

[ ]
sample1.hgwy_mpg.mean()


The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

[ ]
### YOUR WORK HERE

[53]
stats.ttest_1samp(sample1['hgwy_mpg'], 35, nan_policy='omit')


Based on a t-statistic of -3.4 and a p-value of .0076, I reject the null hypothesis that congressional support for the handicapped-infants bill is equal to 50%.






Data Science Unit 1 Sprint 2 Assignment 1
Apply the t-test to real data
Practice 1-Sample T-tests
One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean.

1) Load the Data
Use the automobile dataset
Fix the column headers
Make sure NaNs are used to indicate missing values
Feel free to add code cells and text cells as needed throughout the assignment.

[47]
### YOUR WORK HERE
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt 
import seaborn as sns
from scipy import stats
[48]
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data'
[49]
df = pd.read_csv(url)
[50]
columns = ['symboling','norm_loss','make','fuel','aspiration','doors',
           'bod_style','drv_wheels','eng_loc','wheel_base','length','width',
           'height','curb_weight','engine','cylinders','engine_size',
           'fuel_system','bore','stroke','compression','hp','peak_rpm',
           'city_mpg','hgwy_mpg','price']
[51]
df = pd.read_csv(url, header=0, names=columns)
df.head()

2) Pretend that this dataset represents the cars at a used car lot in your local town.
Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the random_state so that all of us in the class get the same random sample. Please set your random state to 30 when using df.sample()
Is your sample reflective of the population value in regards to highway-mpg? Find the mean for highway-mpg for the entire dataset and compare your sample mean estimate to that value. Can you say that they are different?

[ ]
### YOUR WORK HERE
sample1 = df.sample(10, random_state=30)

[ ]
sample1

[ ]
df.hgwy_mpg.mean()

[ ]
sample1.hgwy_mpg.mean()


The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

[ ]
### YOUR WORK HERE

[53]
stats.ttest_1samp(sample1['hgwy_mpg'], 35, nan_policy='omit')


Based on a t-statistic of -3.4 and a p-value of .0076, I reject the null hypothesis that congressional support for the handicapped-infants bill is equal to 50%.

In [0]:
sample2.hgwy_mpg.mean()

30.53

In [0]:
#still this is not enough to verify claim.

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

The sample with only 10 cars may have a larger deviation to the larger dataset bc of outliers or insuffcient data.

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [0]:
### YOUR WORK HERE
# Grab the file from UCI:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-06-08 22:00:04--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.03s   

2020-06-08 22:00:04 (517 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [0]:
column_headers = ['party','handicapped-infants','water-project',
                  'budget','physician-fee-freeze', 'el-salvador-aid',
                  'religious-groups','anti-satellite-ban',
                  'aid-to-contras','mx-missile','immigration',
                  'synfuels', 'education', 'right-to-sue','crime','duty-free',
                  'south-africa']

df = pd.read_csv('house-votes-84.data', names=column_headers)

In [0]:
df = df.replace({'y': 1, 'n': 0, '?':np.NaN})

df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [0]:
dem = df[df['party'] == 'democrat']
rep = df[df['party'] == 'republican']

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

In [0]:
### YOUR WORK HERE


Null Hypothesis: the two sample means are equal. Republican support this bill (immigration) at the same rate that Democrats do.

Alternative Hypothesis: They support this bill at different rates

Confidence Level: 95%

In [0]:
stats.ttest_ind(rep['right-to-sue'], dem['right-to-sue'], nan_policy='omit')

Ttest_indResult(statistic=13.51064251060933, pvalue=1.2278581709672758e-34)

Conclusion:

Based on a t-statistic of 13.5, and a p-value of 0, we reject the null hypothesis that republicans and democrats support this bill at the same rate, and suggest the alternative, that republicans and democrats show different levels of support for this bill.

In [0]:
print(rep['right-to-sue'].mean())
print(dem['right-to-sue'].mean())

0.8709677419354839
0.14457831325301204


Null Hypothesis: the two sample means are equal. Republican support this bill (budget) at the same rate that Democrats do.

Alternative Hypothesis: They support this bill at different rates

Confidence Level: 95%

In [0]:
stats.ttest_ind(rep['budget'], dem['budget'], nan_policy='omit')

Ttest_indResult(statistic=-23.21277691701378, pvalue=2.0703402795404463e-77)

Conclusion:

Based on a t-statistic of -23.2, and a p-value of 0, we reject the null hypothesis that republicans and democrats support this bill at the same rate, and suggest the alternative, that republicans and democrats show different levels of support for this bill.

In [0]:
print(rep['budget'].mean())
print(dem['budget'].mean())

0.13414634146341464
0.8884615384615384


## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [0]:
### YOUR WORK HERE



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [0]:
### YOUR WORK HERE

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [0]:
### YOUR WORK HERE